toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1388 @@
+  nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback> mErrorCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>
+    mSuccessCallbackHandle;
+
+  PathRunnablesParametersWrapper(
+    const nsAString& aPath,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>&
+      aOnSuccess)
+    : mPath(aPath)
+    , mChangeCallbackHandle(aOnChange)
+    , mErrorCallbackHandle(aOnError)
+    , mSuccessCallbackHandle(aOnSuccess)
+  {
+  }
+};
+
+/**
+ * This runnable is dispatched from the main thread to get the notifications of
+ * the changes in the watched resources by continuously calling the blocking
+ * function GetQueuedCompletionStatus. This function queries the status of the
+ * Completion I/O port initialized in the main thread. The watched resources are
+ * registered to the completion I/O port when calling |addPath|.
+ *
+ * Instead of using a loop within the Run() method, the Runnable reschedules
+ * itself by issuing a NS_DispatchToCurrentThread(this) before exiting. This is
+ * done to allow the execution of other runnables enqueued within the thread
+ * task queue.
+ */
+class NativeFileWatcherIOTask : public Runnable
+{
+public:
+  explicit NativeFileWatcherIOTask()
+    : Runnable("NativeFileWatcherIOTask")
+    , mShuttingDown(false)
+  {
+  }
+
+  NS_IMETHOD Run() override;
+  nsresult AddPathRunnableMethod(
+    PathRunnablesParametersWrapper* aWrappedParameters);
+  nsresult RemovePathRunnableMethod(
+    PathRunnablesParametersWrapper* aWrappedParameters);
+  nsresult DeactivateRunnableMethod();
+
+private:
+  // Maintain 2 indexes - one by resource path, one by resource |HANDLE|.
+  // Since |HANDLE| is basically a typedef to void*, we use nsVoidPtrHashKey to
+  // compute the hashing key. We need 2 indexes in order to quickly look up the
+  // changed resource in the Worker Thread.
+  // The objects are not ref counted and get destroyed by
+  // mWatchedResourcesByPath on NativeFileWatcherService::Destroy or in
+  // NativeFileWatcherService::RemovePath.
+  nsClassHashtable<nsStringHashKey, WatchedResourceDescriptor>
+    mWatchedResourcesByPath;
+  nsDataHashtable<nsVoidPtrHashKey, WatchedResourceDescriptor*>
+    mWatchedResourcesByHandle;
+
+  // The same callback can be associated to multiple watches so we need to keep
+  // them alive as long as there is a watch using them. We create two hashtables
+  // to map directory names to lists of nsMainThreadPtr<callbacks>.
+  nsClassHashtable<nsStringHashKey, ChangeCallbackArray> mChangeCallbacksTable;
+  nsClassHashtable<nsStringHashKey, ErrorCallbackArray> mErrorCallbacksTable;
+
+  // Other methods need to know that a shutdown is in progress.
+  bool mShuttingDown;
+
+  // Main inotify file descriptor
+  int mInotifyFileDescriptor;
```

Comment:
Here, too, either make it clear that it is auto-closed (using `Scoped`) or document where it is closed. Can all threads close it?